lending_pool/src/factory.rs

```diff
@@ -26,20 +29,59 @@ pub trait FactoryModule {
         arg_buffer.push_arg(r_slope2);
         arg_buffer.push_arg(u_optimal);
         arg_buffer.push_arg(reserve_factor);
+        arg_buffer.push_arg(liquidation_threshold);
 
-        let (address, _) = self.raw_vm_api().deploy_contract(
+        let (new_address, _) = self.raw_vm_api().deploy_from_source_contract(
             self.blockchain().get_gas_left(),
-            &amount,
-            bytecode,
-            code_metadata,
+            &BigUint::zero(),
+            &self.liq_pool_template_address().get(),
+            CodeMetadata::UPGRADEABLE,
             &arg_buffer,
         );
 
-        address
+        Ok(new_address)
     }
 
-    // can be implemented when upgrade is available in elrond-wasm
-    fn upgrade_pool(&self, _pool_address: &Address, _new_bytecode: &ManagedBuffer) -> bool {
-        true
+    fn upgrade_pool(
+        &self,
+        lp_address: ManagedAddress,
+        base_asset: &TokenIdentifier,
+        lending_pool_address: &ManagedAddress,
+        r_base: BigUint,
+        r_slope1: BigUint,
+        r_slope2: BigUint,
+        u_optimal: BigUint,
+        reserve_factor: BigUint,
+        liquidation_threshold: BigUint,
+    ) -> SCResult<()> {
+        require!(
+            !self.liq_pool_template_address().is_empty(),
+            "liquidity pool contract template is empty"
+        );
+
+        let mut arg_buffer = ManagedArgBuffer::new_empty(self.type_manager());
+        arg_buffer.push_arg(base_asset);
+        arg_buffer.push_arg(lending_pool_address);
+        arg_buffer.push_arg(r_base);
+        arg_buffer.push_arg(r_slope1);
+        arg_buffer.push_arg(r_slope2);
+        arg_buffer.push_arg(u_optimal);
+        arg_buffer.push_arg(reserve_factor);
+        arg_buffer.push_arg(liquidation_threshold);
+
+        self.raw_vm_api().upgrade_from_source_contract(
+            &lp_address,
+            self.blockchain().get_gas_left(),
+            &BigUint::zero(),
+            &self.liq_pool_template_address().get(),
+            CodeMetadata::UPGRADEABLE,
+            &arg_buffer,
+        );
+
+        Ok(())
```

Comment:
Use the same proxy you used for deploy, simply change `deploy_from_source` with `upgrade_from_source`, and pass the SC address as argument instead of `ManagedAddress::zero()`